### PR TITLE
feat: add group and projects to resource pool authz

### PR DESCRIPTION
**Pull request by SalimKayal**
## Summary

This PR extends resource pool authorization to support **group-level and project-level membership**, in addition to individual users. It introduces a new Authzed schema (v10) with `group_viewer` and `project_viewer` relations, refactors the CRC user repository into a polymorphic `MemberRepository`, and adds comprehensive tests for inheritance, prohibition, and edge cases.

## Motivation

Previously, resource pools could only be shared with individual users via direct `viewer` relations or blocked via `prohibited`. This did not scale for organizations that manage access through Renku groups or projects. By allowing groups and projects as first-class members of resource pools, access control becomes declarative and maintainable at the team level.

### Authz / SpiceDB Schema
- **New schema v10** (`components/renku_data_services/authz/schemas.py`):
  - Added `group_viewer` and `project_viewer` relations on `resource_pool`.
  - `read` permission now traverses `group_viewer->direct_member` and `project_viewer->direct_member` to resolve inherited access.
  - Preserved existing `viewer`, `public_viewer`, `prohibited`, and `resource_pool_platform->is_admin` semantics.
- **Migration** (`migrations/versions/65f8330c4d25_authz_migrate_to_v10.py`):
  - Upgrades/downgrades Authzed schema with proper relationship cleanup on rollback.

### Domain Models
- Added `MemberType` enum and `ResourcePoolMemberIdentifier` dataclass (`crc/models.py`) to represent polymorphic members (USER, GROUP, PROJECT) with ULID validation for non-user types.
- Extended `Member` model with `subject_type` field (`authz/models.py`) so membership changes carry the target resource type.
- Added `Member.with_resource_pool()` factory for building resource-pool-specific members.

### Repository Refactor
- Renamed `UserRepository` → `MemberRepository` (`crc/db.py`) to reflect its broader scope.
- Injected `GroupRepository` and `ProjectRepository` into `MemberRepository` for resolving group/project identifiers.
- Refactored internal methods:
  - `_grant_resource_pool_members`, `_revoke_resource_pool_members`, `_prohibit_resource_pool_users`, `_unprohibit_resource_pool_users` now accept `ResourcePoolMemberIdentifier` collections.
  - Added `_resolve_members()` to validate and map members to `(id, ResourceType)` tuples, rejecting missing groups/projects with `MissingResourceError`.
- Added public `grant_resource_pool_members` / `revoke_resource_pool_members` admin-only methods.

### Dependency Wiring
- Updated `DependencyManager`, blueprints, and app registration to use `member_repo` instead of `user_repo` throughout `data_api`.

### Supporting Queries
- Added `GroupRepository.get_group_by_id()` (used internally for member resolution).
- Added `ProjectRepository.get_project_by_id()` (admin-only, for resolving project references back to human-readable paths).

### Tests
- **Schema tests** (`test_schemas.py`): Full v10 fixture with `assertTrue`/`assertFalse` and validation paths for public, private, group-inherited, and project-inherited pools.
- **Authorization tests** (`test_authorization.py`):
  - `test_resource_pool_member_relationships_subject_types`: verifies correct subject types in authz updates.
  - `test_resource_pool_polymorphic_inheritance`: confirms group/project members gain access via `direct_member`.
  - `test_resource_pool_polymorphic_prohibition`: confirms `prohibited` still blocks regardless of membership type.
  - `test_resource_pool_admin_bypass_polymorphic`: confirms admin bypass.
  - `test_resource_pool_no_indirect_inheritance`: confirms only `direct_member` is traversed (not namespace owners).

## Backward Compatibility

- Existing user-only pools continue to work unchanged.
- The old `viewer` relation on `resource_pool` is preserved; only `group_viewer` and `project_viewer` are new.
- The DB migration is reversible (downgrade restores v9 and deletes v10 resource_pool relationships).
- **Warning**: Downgrading removes the group_viewer and project_viewer relationship, revoking access to the resource pool.

## Checklist

- [x] Authzed schema v10 defined and migrated
- [x] `MemberRepository` supports user/group/project identifiers
- [x] Admin-only grant/revoke API surface ready
- [x] SpiceDB schema tests pass for inheritance and prohibition
- [x] Authorization integration tests pass

## PR Stack

* base: #1311 
  * [this](#1267)
    * #1299 
  
/deploy

### CODE REVIEW

File: components/renku_data_services/authz/authz.py [hunk 1]
```diff
@@ -187,7 +193,7 @@ class AuthzOperation(StrEnum):
 
 class _AuthzConverter:
     @staticmethod
-    def project(id: ULID) -> ObjectReference:
+    def project(id: str | ULID) -> ObjectReference:
```
Comment (leafty) [inline components/renku_data_services/authz/authz.py]: I don't think we should add a lax type here (and below). Why would we ever have a non-ULID project ID?

File: components/renku_data_services/crc/models.py [hunk 2]
```diff
@@ -487,3 +489,82 @@ class ResourcePoolMembershipChange:
...
+    """Identifier for a member in a resource pool.
+
+    Identifier semantics by member_type:
+      - USER    : Keycloak user id (string)
+      - GROUP   : group slug, e.g. "my-group"
+      - PROJECT : project path slug, e.g. "my-group/my-project"
+
+    Groups and projects are addressed by their human-readable slug rather than
+    by internal id (ULID). This mirrors how they are referenced elsewhere in the
+    public API and avoids leaking storage identifiers into admin workflows.
```
Comment (leafty) [inline components/renku_data_services/crc/models.py]: What? Why? Use ULIDs. The ULIDs are not secret nor meant to be hidden. Also, slugs can be modified by users (change slug, move project to a different group, etc.).

**Review by leafty — CHANGES_REQUESTED** [review]

File: components/renku_data_services/crc/models.py (hunk 2 above)
Comment (leafty) [reply components/renku_data_services/crc/models.py]: Also, see how data connectors are linked to projects in the API: ULIDs for both sides (projects and data connectors) so that these links are stable through slug changes.

Comment (SalimKayal) [reply components/renku_data_services/crc/models.py]: noted, will change that.

File: components/renku_data_services/authz/authz.py (hunk 1 above)
Comment (SalimKayal) [reply components/renku_data_services/authz/authz.py]: reverted these.

File: components/renku_data_services/crc/models.py (hunk 2 above)
Comment (SalimKayal) [reply components/renku_data_services/crc/models.py]: done

**Review by leafty — APPROVED** [review]
LGTM
